lumen/ai/tools.py

```diff
@@ -104,45 +104,100 @@ class TableLookup(VectorLookupTool):
 
+    include_metadata = param.Boolean(default=True, doc="""
+        Whether to include table descriptions in the embeddings and responses.""")
+
+    include_columns = param.Boolean(default=False, doc="""
+        Whether to include column descriptions in the embeddings.""")
+
+    max_concurrent = param.Integer(default=5, doc="""
+        Maximum number of concurrent metadata fetch operations.""")
+
     def __init__(self, **params):
         super().__init__(**params)
+        self._table_metadata = {}
+        self._metadata_tasks = set()  # Track ongoing metadata fetch tasks
+        self._semaphore = asyncio.Semaphore(self.max_concurrent)
         self._memory.on_change('sources', self._update_vector_store)
+
         if sources := self._memory.get("sources"):
             self._update_vector_store(None, None, sources)
         elif source := self._memory.get("source"):
             self._update_vector_store(None, None, [source])
 
+    async def _fetch_and_store_metadata(self, source, table):
+        """Fetch metadata for a table and store it in the vector store."""
+        async with self._semaphore:
+            metadata = await asyncio.to_thread(source.get_metadata, table)
+        multi_source = len(self._memory.get("sources", [])) > 1
+        if multi_source:
+            enriched_text = table_name = f"{source.name}{SOURCE_TABLE_SEPARATOR}{table}"
+        else:
+            enriched_text = table_name = table
+        self._table_metadata[table_name] = metadata
+
+        vector_metadata = {"table": table, "source": source.name}
+        if self.vector_store.filter_by(vector_metadata):
+            return
+
+        if description := metadata.get('description', ''):
+            enriched_text += f"\nDescription: {description}"
+        if self.include_columns and (columns := metadata.get('columns', {})):
+            enriched_text += "\nColumns:"
+            for col_name, col_info in columns.items():
+                col_text = f"\n- {col_name}"
+                if 'description' in col_info:
+                    col_text += f": {col_info['description']}"
+                enriched_text += col_text
+        self.vector_store.add([{"text": enriched_text, "metadata": vector_metadata}])
+
     def _update_vector_store(self, _, __, sources):
         for source in sources:
             for table in source.get_tables():
-                source_table = f'{source.name}{SOURCE_TABLE_SEPARATOR}{table}'
-                if not self.vector_store.query(source_table, threshold=1):
-                    self.vector_store.add([{"text": source_table}])
+                metadata = {"table": table, "source": source.name}
+                if not self.include_metadata:
+                    if self.vector_store.filter_by(metadata):
+                        # TODO: Add ability to update existing source table?
+                        continue
+                    source_table = f"{source.name}{SOURCE_TABLE_SEPARATOR}{table}"
```

Comment:
Maybe this should be renamed to table_name, or better yet, let's officially differentiate "table_name" as "table_slug" (so we know it's attached to source)